# **Project 2 - Sales Analysis - ETL**

## Objectives

- Extract data from provided CSV files
- Clean it 
- Apply feature engineering if necessary 
- Remove any unnecessary columns
- Save to a new CSV file


## Inputs

CSV files provided:

stores data-set.csv
sales data-set.csv
features data set.csv

Renamed files to a uniform naming convention:

Sales_Features_DataSet.csv
Sales_DataSet.csv
Sales_Stores_DataSet.csv

Note: original files are stored in Data/OriginalFiles


## Outputs

CSV file created from ETL etc stored in Data/CleanedDataSets:

Sales_Stores_DataSet_Cleaned.csv


## Additional Comments

Developed an experimental ETL library which is in this project (modETL_library.py) 
Has lots of cool features so will be interesting to see how it works "in the field"

Used an AI tool to help with the ETL process:
- GitHub 
- Copilot 

See Documents/What_AI_Used_For.md for more details.

## Initalise Working Environment

In [1]:
#import libraries
import os
import numpy as np
import pandas as pd

#below solution provided by chatGPT 
import sys
from pathlib import Path

project_root = Path.cwd().parent

if str(project_root / "assets" / "python_files") not in sys.path:
    sys.path.insert(0, str(project_root / "assets" / "python_files"))
    
import modGlobal
import modETL_Library as modETL
#end solution provided by chatGPT

# Section 1 - Initalisation

## Intitialise All Variables To Be Used In Global Stack 

In [ ]:

#DataFrame vars for ETL
dfSales_Stores_DataSet = None
dfSales_Stores_DataSet_Work = None
dfCleaned = None

#missing values check vars
intLessThanZero = 0

#stores current directory
strCurrentDir = ""

#other vars
dictTypeMapping = {}
dictDataFrames = dict()
lstColumns = list()

## Set Current Directory To Base Project Directory

In [3]:
#get project directory - default is jupyter notebook sub folder as that is where this file is located!
#so move back one to the project root path
# Source - https://stackoverflow.com/a/17726833
# Posted by chimpsarehungry
# Retrieved 2026-07-05, License - CC BY-SA 3.0

#get current folder
strCurrentDir = os.getcwd()

#is the last part of the path the project directory?
if not strCurrentDir.endswith(modGlobal.CNST_STR_PROJECT_DIR):
   #get current working directory and move back one to the project root path
   strCurrentDir =  os.path.normpath(os.getcwd() + os.sep + os.pardir)
   os.chdir(os.path.dirname(strCurrentDir))
   #change directory
   os.chdir(strCurrentDir)

#confirm current directory is project directory
print(f"Current Directory: \n {os.getcwd()}")

Current Directory: 
 /Users/rogerwilliams/Projects/Python/CourseProjects/Project2-SalesAnalysis


# Section 1 - Extraction

- Read csv file sales data-set.csv as other two do not require ETL (boo!)
- Move into working files directory
- Read csv file into a pandas dataframe
- Get schema info -> column and row numbers
- Get first 5 records
- Get list of column datatypes
- Get detailed schema infomation

## Read csv File Into Variable For Processing

In [4]:
#read csv file into DataFrame
#now repeat for the other files
dfSales_Stores_DataSet = modETL.funcReadFileReturnDataFrame("Stores_DataSet.csv")
#save as working csv file
modETL.funcSaveDataFrameToWorkingFile(dfSales_Stores_DataSet)

#read file from working files folder
#ETL library returns a dictionary of all files in the folder with the attribute name
#set to the actual csv filename
dictDataFrames = modETL.funcReadWorkingFilesReturnDictionary()
dfSales_Stores_DataSet = dictDataFrames["Stores_DataSet_Working.csv"]

Folder Structure Created Successfully!
Read Stores_DataSet.csv Into DataFrame

Saved: Stores_DataSet To Working Folder
3 csv Files Read Into DataFrames

DataFrames Created:
Features_DataSet_Working.csv
Stores_DataSet_Working.csv
Sales_DataSet_Working.csv




## Get Schema Info - Sales_Stores_DataSet

In [5]:
modETL.funcGetStructure(dfSales_Stores_DataSet)
dfSales_Stores_DataSet.head()

DataFrame Structure:
<class 'pandas.DataFrame'>
RangeIndex: 45 entries, 0 to 44
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   Unnamed: 0  45 non-null     int64
 1   Store       45 non-null     int64
 2   Type        45 non-null     str  
 3   Size        45 non-null     int64
dtypes: int64(3), str(1)
memory usage: 1.6 KB
None


Summary of DataFrame Structure:
Numeric Columns:

       Unnamed: 0      Store           Size
count   45.000000  45.000000      45.000000
mean    22.000000  23.000000  130287.600000
std     13.133926  13.133926   63825.271991
min      0.000000   1.000000   34875.000000
25%     11.000000  12.000000   70713.000000
50%     22.000000  23.000000  126512.000000
75%     33.000000  34.000000  202307.000000
max     44.000000  45.000000  219622.000000


String Columns:

       Type
count    45
unique    3
top       A
freq     22


No DateTime   Columns Found!




,Unnamed: 0,Store,Type,Size
0,0,1,A,151315
1,1,2,A,202307
2,2,3,B,37392
3,3,4,A,205863
4,4,5,B,34875


## Observations - Sales_Stores_DataSet
45 rows
4 columns

Columns:

Unnamed, Store, Type, Size

Assuming Unnamed is an index column and can be dropped if necessary

We can also see the data types:

Unnamed : int64
Store : int64
Type : str
Size : int64

In the Summary of Dataframe Structure for numerical values we see this:

DataFrame Structure:
===================
<class 'pandas.DataFrame'>. 
RangeIndex: 45 entries, 0 to 44   
Data columns (total 4 columns):  
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  45 non-null     int64  
 1   Store       45 non-null     int64  
 2   Type        45 non-null     str    
 3   Size        45 non-null     int64  
dtypes: int64(3), str(1)  
memory usage: 1.5 KB  
None  

Summary of DataFrame Structure:
===============================
Numeric Columns:

       Unnamed: 0      Store           Size  
count   45.000000  45.000000      45.000000  
mean    22.000000  23.000000  130287.600000  
std     13.133926  13.133926   63825.271991  
min      0.000000   1.000000   34875.000000  
25%     11.000000  12.000000   70713.000000  
50%     22.000000  23.000000  126512.000000  
75%     33.000000  34.000000  202307.000000  
max     44.000000  45.000000  219622.000000  

**What Does This Mean?**

- Count is how many values are in the columns
- Mean is the average value of the column
- Std is the standard deviation of the column 
  Standard deviation is how far from the average (mean) the values are. The closer to 0 the
  more consistent the values are i.e. not much deviation from the mean. The deviation range 
  can be referred to as the Sigma. The measurement is based on the Alpha this is a percentage
  of the mean used to determine variance. ypically this is set to 5%
- Min is the minimum value of the column
- 25% is the value at the 25th percentile of the column
- 50% is the value between the 25th and 75th percentiles of the column
- 75% is the value at the 75th percentile of the column

The middle 50% is also called the Interquartile Range (IQR) and its width can be set by changing the
alpha value which is normally 5%

**What?**

Percentile is a technical term for a 25% segment of the data. 
50% is the mean, 25% is below that and 75% is 25%above that, this crude picture illustrates this:

 1st Percentile      2nd Percentile        3rd Percentile
0      -      25%  25%      -     75%   75%      -     100%  

Note: 2nd percentile can also be called Central Tendency

- Max is the maximum value of the column

String Columns:

       Type
count    45 - number of records  
unique    3 - number of unique values  
top       A - most frequent value  
freq     22 - frequency of most frequent value


# Get Column Data Types

In [6]:
#get column data types
print (f"Data Types: \n{dfSales_Stores_DataSet.dtypes}")

Data Types: 
Unnamed: 0    int64
Store         int64
Type            str
Size          int64
dtype: object


## Observations - Sales_Stores_DataSet

Type is a string - needs conversion to datetime  
Other column types look ok
If unnamed causes issues later on will drop the column!

Observations:
- No missing values
- High duplicates but this is to be expected as type would be repeated

Column: Type looks like a candidate for feature engineering


## Get Schema Statistics

Look for duplicates and missing values

In [7]:
#get column statistics
modETL.funcGetStatistics(dfSales_Stores_DataSet)

DataFrame Statistics:
       Unnamed: 0      Store           Size
count   45.000000  45.000000      45.000000
mean    22.000000  23.000000  130287.600000
std     13.133926  13.133926   63825.271991
min      0.000000   1.000000   34875.000000
25%     11.000000  12.000000   70713.000000
50%     22.000000  23.000000  126512.000000
75%     33.000000  34.000000  202307.000000
max     44.000000  45.000000  219622.000000


DataFrame Shape:
Rows: 45, Columns: 4


Missing Values Per Column:


Duplicate Value Count Per Column:
Type                 Total Values: 45  Duplicate Values: 42, 93.33% Percent Duplicate Values
Size                 Total Values: 45  Duplicate Values: 5, 11.11% Percent Duplicate Values



## Observations - Sales_Stores_DataSet

Type - is a string and a candidate for feature engineering  
Other column types look ok
Store does not appear as is nothing but unique values!

Observations:
- No missing values
- High duplicates but this is to be expected as type would be repeated


## Get Unique Values - Sales_Stores_DataSet

In [8]:
#show unique values count
modETL.funcGetUniqueValuesCount(dfSales_Stores_DataSet)

DataFrame Unique Values Per Column:
Unnamed: 0           - 45: Unique Values Out Of 45 Total Values
Store                - 45: Unique Values Out Of 45 Total Values
Type                 - 3: Unique Values Out Of 45 Total Values
Size                 - 40: Unique Values Out Of 45 Total Values



# Observations - Sales_Stores_DataSet

Unnamed has 45 unique values (because it is an index column)  
Store has 45 unique values as there are 45 stores in the dataset  
Type has 3 unique values (A, B, C) - this is a candidate for possible feature engineering  
Size has 45 unique values as each store has a different size  

## Observations - Sales_Stores_DataSet

Nothing exciting here but is always worth a look!

---

## Observations - Sales_DataSet

Store has 45 unique values  
Type has 3 unique values  
Size has 40 unique values  


## Get Categorical Value Distribution

In [9]:
#get categorical distribution for object columns (non numeric)
for objColumn in ["Type"]:
    print( f"[{objColumn}] Value Distribution:")
    print( dfSales_Stores_DataSet[objColumn].value_counts())

[Type] Value Distribution:
Type
A    22
B    17
C     6
Name: count, dtype: int64


## Observations - Sales_DataSet

No deviations as the column data is just A or B or C


# Section 2

- If missing values determine what to fill with (mean, mode or categorical something else)
- Transform data types if necessary


In [10]:
## Create Copy Of DataFrame
dfSales_Stores_DataSet_Work = dfSales_Stores_DataSet.copy()

## Strip Spaces From Type Column

In [11]:
#remove spaces from type column (string datatype)
for objColumn in["Type"]:
    dfSales_Stores_DataSet_Work[objColumn] = dfSales_Stores_DataSet_Work[objColumn].str.strip()
    
#show first 5 rows to check    
dfSales_Stores_DataSet_Work.head()    

,Unnamed: 0,Store,Type,Size
0,0,1,A,151315
1,1,2,A,202307
2,2,3,B,37392
3,3,4,A,205863
4,4,5,B,34875


## Observations - Sales_Stores_DataSet

No data corruptions

## Validate Numerical Values

see if Stores column has values less than 1
see if Dept column has values less than 1
see if Weekly_Sales column have values less then 0

In [12]:
#get count of how many values in Stores column are less than 1
intLessThanOne = (dfSales_Stores_DataSet_Work["Store"] < 1).sum()
#print result
intLessThanOne

np.int64(0)

---

## Observations - Sales_Stores_DataSet

Values for Store in acceptable limits

## Validate Size Values

In [13]:
#get count of how many values in Size column are less than 1
intLessThanOne = (dfSales_Stores_DataSet_Work["Size"] < 1).sum()
#print result
intLessThanOne

np.int64(0)

## Observations - Sales_Stores_DataSet

Values for Size in acceptable limits

## Data Transformations/Feature Engineering

Create column: Store_Type as integer mapped from existing Type column  
i.e. have 1, 2, 3 values instead of Type columns values A, B, C  
  
Doing here as much smaller dataset than the one being used for visualisations  
which will be a combination of all 3 csv files.  


In [14]:
#create column:StoreType as integer 
#Copilot generated code applied naming convention

dictTypeMapping = {"A": 1, "B": 2, "C": 3} 
dfSales_Stores_DataSet_Work["Store_Type"] = dfSales_Stores_DataSet_Work["Type"].map(dictTypeMapping).astype("int64")

#check results get first 50 rows
dfSales_Stores_DataSet_Work.head(50)


,Unnamed: 0,Store,Type,Size,Store_Type
0,0,1,A,151315,1
1,1,2,A,202307,1
2,2,3,B,37392,2
3,3,4,A,205863,1
4,4,5,B,34875,2
5,5,6,A,202505,1
6,6,7,B,70713,2
7,7,8,A,155078,1
8,8,9,B,125833,2
9,9,10,B,126512,2


## Observations - Sales_Stores_DataSet

Column created successfully, data looks good

## Check Data Type For New Column Is Correct

In [15]:
#get schema data types
dfSales_Stores_DataSet_Work.dtypes

Unnamed: 0    int64
Store         int64
Type            str
Size          int64
Store_Type    int64
dtype: object

## Observations - Sales_DataSet

Data type for new column StoreType is correct

## Remove Unnecessary Columns

In [16]:
#nothing to do as no viable columns to remove!


## Observations - Sales_Stores_DataSet

Column created successfully, data looks good

## Save Work DataFrame To CSV File

In [17]:
#create list of columns new and old JUST what is actually needed
lstColumns = [
                #original columns
                "Store","Size",
                #new columns
                "Store_Type"
              ]

#create new DataFrame for cleaned data
dfCleaned = dfSales_Stores_DataSet_Work[lstColumns].copy()

modETL.funcSaveDataFrameToCleanedFile(dfCleaned)

#Note: excluded column: Dept as not in the hypothesis analysis which is a nice way
#      avoiding the fact I don't know just how to plot a chart with 98 departments
#      and have it readable!

Saved: Stores_DataSet To Visualisation Folder


# Conclusions and Next Steps

ETL went smoothly, DataSet ready for use